From 7f1e4a4d78fab8c8ef8ec6911452f535d853f738 Mon Sep 17 00:00:00 2001
From: GitHub Actions <actions@github.com>
Date: Sun, 4 Oct 2026 09:41:14 +0000
Subject: [PATCH] Email-Fix: Don't fail hard on send email errors.

- Add try catch environment to send email commands, so errors are correctly logged.
- Add environment flag APP_EMAIL_ENABLED
---
 app/Jobs/Emails/EmailChangeJob.php            | 24 +++++++++---
 app/Jobs/Emails/Purge/FirstWarningJob.php     | 38 +++++++++++--------
 app/Jobs/Emails/Purge/SecondWarningJob.php    | 36 ++++++++++--------
 .../Emails/SubscriptionCancelEmailJob.php     | 18 ++++++++-
 .../Emails/SubscriptionFailedEmailJob.php     | 25 +++++++++---
 .../Subscriptions/EmailValidationJob.php      | 24 +++++++++---
 .../Subscriptions/PaypalExpiringAlert.php     | 20 ++++++++--
 .../Subscriptions/UpcomingYearlyAlert.php     | 25 +++++++++---
 .../WelcomeSubscriptionEmailJob.php           | 23 +++++++++--
 app/Jobs/Emails/WelcomeEmailJob.php           | 32 +++++++++-------
 app/Jobs/Users/NewPassword.php                | 32 +++++++++-------
 config/app.php                                |  1 +
 12 files changed, 211 insertions(+), 87 deletions(-)

diff --git a/app/Jobs/Emails/EmailChangeJob.php b/app/Jobs/Emails/EmailChangeJob.php
index 9e10a32a4..841cc1018 100644
--- a/app/Jobs/Emails/EmailChangeJob.php
+++ b/app/Jobs/Emails/EmailChangeJob.php
@@ -44,10 +44,24 @@ class EmailChangeJob implements ShouldQueue
         }
 
         // Send an email to the user
-        Mail::to($this->email)
-            ->locale($user->locale)
-            ->send(
-                new EmailChangeMail($user)
-            );
+        if ( config('app.email_enabled') )
+        {
+            try {
+              Mail::to($this->email)
+                  ->locale($user->locale)
+                  ->send(
+                      new EmailChangeMail($user)
+                  );
+            } catch (\GuzzleHttp\Exception\ServerException $e) {
+                // Silence
+            } catch (TransportExceptionInterface $e) {
+                // Mail isn't configured or the server is unreachable. This is allowed, so don't rethrow.
+                Log::error('Change email email not sent: ' . $e->getMessage());
+            } catch (Exception $e) {
+                // Something went wrong with mailgun, or the email is invalid. Silence these errors
+                // to avoid spamming sentry.
+                Log::error('Change email email not sent: ' . $e->getMessage());
+            }
+        }
     }
 }
diff --git a/app/Jobs/Emails/Purge/FirstWarningJob.php b/app/Jobs/Emails/Purge/FirstWarningJob.php
index d24e59dea..c28efbc67 100644
--- a/app/Jobs/Emails/Purge/FirstWarningJob.php
+++ b/app/Jobs/Emails/Purge/FirstWarningJob.php
@@ -54,22 +54,28 @@ class FirstWarningJob implements ShouldQueue
         $campaigns = $service->user($user)->flaggedCampaigns();
         UserLogger::user($user)->log(UserAction::purgeWarningFirst);
 
-        $target = app()->isProduction() ? $user->email : config('mail.from.address');
-        if (empty($target)) {
-            return;
-        }
-        try {
-            Mail::to($target)
-                ->locale($user->locale ?? 'en-US')
-                ->send(
-                    new FirstWarning($user, $campaigns)
-                );
-        } catch (ServerException $e) {
-            // Silence
-        } catch (Exception $e) {
-            // Something went wrong with mailgun, or the email is invalid. Silence these errors
-            // to avoid spamming sentry.
-            throw $e;
+        if ( config('app.email_enabled') ) 
+        {
+            $target = app()->isProduction() ? $user->email : config('mail.from.address');
+            if (empty($target)) {
+                return;
+            }
+            try {
+                Mail::to($target)
+                    ->locale($user->locale ?? 'en-US')
+                    ->send(
+                        new FirstWarning($user, $campaigns)
+                    );
+            } catch (ServerException $e) {
+                // Silence
+            } catch (TransportExceptionInterface $e) {
+                // Mail isn't configured or the server is unreachable. This is allowed, so don't rethrow.
+                Log::debug('First purge warning email not sent: ' . $e->getMessage());
+            } catch (Exception $e) {
+                // Something went wrong with mailgun, or the email is invalid. Silence these errors
+                // to avoid spamming sentry.
+                Log::error('First purge warning email not sent: ' . $e->getMessage());
+            }
         }
     }
 }
diff --git a/app/Jobs/Emails/Purge/SecondWarningJob.php b/app/Jobs/Emails/Purge/SecondWarningJob.php
index c41dce4f1..87d968e2f 100644
--- a/app/Jobs/Emails/Purge/SecondWarningJob.php
+++ b/app/Jobs/Emails/Purge/SecondWarningJob.php
@@ -54,22 +54,26 @@ class SecondWarningJob implements ShouldQueue
 
         UserLogger::user($user)->log(UserAction::purgeWarningSecond);
 
-        $target = app()->isProduction() ? $user->email : config('mail.from.address');
-        if (empty($target)) {
-            return;
-        }
-        try {
-            Mail::to($target)
-                ->locale($user->locale ?? 'en-US')
-                ->send(
-                    new SecondWarning($user, $campaigns)
-                );
-        } catch (ServerException $e) {
-            // Silence
-        } catch (Exception $e) {
-            // Something went wrong with mailgun, or the email is invalid. Silence these errors
-            // to avoid spamming sentry.
-            throw $e;
+        if ( config('app.email_enabled') )
+        {
+            $target = app()->isProduction() ? $user->email : config('mail.from.address');
+            if (empty($target)) {
+                return;
+            }
+            try {
+                Mail::to($target)
+                    ->locale($user->locale ?? 'en-US')
+                    ->send(
+                        new SecondWarning($user, $campaigns)
+                    );
+            } catch (ServerException $e) {
+                // Silence
+                Log::debug('Second warning email not sent: ' . $e->getMessage());
+            } catch (Exception $e) {
+                // Something went wrong with mailgun, or the email is invalid. Silence these errors
+                // to avoid spamming sentry.
+                Log::error('Second warning email not sent: ' . $e->getMessage());
+            }
         }
     }
 }
diff --git a/app/Jobs/Emails/SubscriptionCancelEmailJob.php b/app/Jobs/Emails/SubscriptionCancelEmailJob.php
index f4bf58824..fb78b6147 100644
--- a/app/Jobs/Emails/SubscriptionCancelEmailJob.php
+++ b/app/Jobs/Emails/SubscriptionCancelEmailJob.php
@@ -48,8 +48,22 @@ class SubscriptionCancelEmailJob implements ShouldQueue
         //    ->send(new CancelledSubscriptionMail($cancellation));
 
         // Send an email to the user
-        Mail::to($user->email)
-            ->send(new CancelledUserSubscriptionMail($cancellation));
+        if ( config('app.email_enabled') )
+        {
+            try {
+                Mail::to($user->email)
+                    ->send(new CancelledUserSubscriptionMail($cancellation));
+            } catch (\GuzzleHttp\Exception\ServerException $e) {
+                // Silence
+            } catch (TransportExceptionInterface $e) {
+                // Mail isn't configured or the server is unreachable. This is allowed, so don't rethrow.
+                Log::warning('Cancel subscription email not send: ' . $e->getMessage());
+            } catch (Exception $e) {
+                // Something went wrong with mailgun, or the email is invalid. Silence these errors
+                // to avoid spamming sentry.
+                Log::error('Cancel subscription email not send: ' . $e->getMessage());
+            }
+        }
 
         UserLogger::user($user)->log(UserAction::subCancelManual);
     }
diff --git a/app/Jobs/Emails/SubscriptionFailedEmailJob.php b/app/Jobs/Emails/SubscriptionFailedEmailJob.php
index a835e5825..b471d344c 100644
--- a/app/Jobs/Emails/SubscriptionFailedEmailJob.php
+++ b/app/Jobs/Emails/SubscriptionFailedEmailJob.php
@@ -60,11 +60,26 @@ class SubscriptionFailedEmailJob implements ShouldQueue
             );*/
 
         // Send an email to the user
-        Mail::to($user->email)
-            // ->bcc('hello@kanka.io')
-            ->send(
-                new FailedUserSubscriptionMail($user)
-            );
+        if ( config('app.email_enabled') )
+        {
+            try {
+                Mail::to($user->email)
+                    // ->bcc('hello@kanka.io')
+                    ->send(
+                        new FailedUserSubscriptionMail($user)
+                    );
+            } catch (\GuzzleHttp\Exception\ServerException $e) {
+                // Silence
+            } catch (TransportExceptionInterface $e) {
+                // Mail isn't configured or the server is unreachable. This is allowed, so don't rethrow.
+                Log::debug('Failed subscription email not send: ' . $e->getMessage());
+            } catch (Exception $e) {
+                // Something went wrong with mailgun, or the email is invalid. Silence these errors
+                // to avoid spamming sentry.
+                Log::error('Failed subscription email not sent: ' . $e->getMessage());
+            }
+        }
+
         UserLogger::user($user)->log(UserAction::failedChargeEmail);
     }
 }
diff --git a/app/Jobs/Emails/Subscriptions/EmailValidationJob.php b/app/Jobs/Emails/Subscriptions/EmailValidationJob.php
index c3509e6f8..48d747c9a 100644
--- a/app/Jobs/Emails/Subscriptions/EmailValidationJob.php
+++ b/app/Jobs/Emails/Subscriptions/EmailValidationJob.php
@@ -44,10 +44,24 @@ class EmailValidationJob implements ShouldQueue
         $userValidation = UserValidation::find($this->token);
         $url = route('validation.email', ['userValidation' => $userValidation]);
 
-        Mail::to($user->email)
-            ->locale($user->locale)
-            ->send(
-                new ValidationEmail($user, $url)
-            );
+        if ( config('app.email_enabled') )
+        {
+            try {
+                Mail::to($user->email)
+                    ->locale($user->locale)
+                    ->send(
+                        new ValidationEmail($user, $url)
+                    );
+            } catch (\GuzzleHttp\Exception\ServerException $e) {
+                // Silence
+            } catch (TransportExceptionInterface $e) {
+                // Mail isn't configured or the server is unreachable. This is allowed, so don't rethrow.
+                Log::warning('Email validation email not sent: ' . $e->getMessage());
+            } catch (Exception $e) {
+                // Something went wrong with mailgun, or the email is invalid. Silence these errors
+                // to avoid spamming sentry.
+                Log::error('Email validation email not sent: ' . $e->getMessage());
+            }
+        }
     }
 }
diff --git a/app/Jobs/Emails/Subscriptions/PaypalExpiringAlert.php b/app/Jobs/Emails/Subscriptions/PaypalExpiringAlert.php
index ba0fdee01..3a23eee03 100644
--- a/app/Jobs/Emails/Subscriptions/PaypalExpiringAlert.php
+++ b/app/Jobs/Emails/Subscriptions/PaypalExpiringAlert.php
@@ -35,9 +35,23 @@ class PaypalExpiringAlert implements ShouldQueue
             return;
         }
 
-        Mail::to($user->email)
-            ->locale($user->locale)
-            ->send(new PaypalExpiringMail($user));
+        if ( config('app.email_enabled') )
+        {
+            try {
+                Mail::to($user->email)
+                    ->locale($user->locale)
+                    ->send(new PaypalExpiringMail($user));
+            } catch (\GuzzleHttp\Exception\ServerException $e) {
+                // Silence
+            } catch (TransportExceptionInterface $e) {
+                // Mail isn't configured or the server is unreachable. This is allowed, so don't rethrow.
+                Log::warning('Paypal expired email not sent: ' . $e->getMessage());
+            } catch (Exception $e) {
+                // Something went wrong with mailgun, or the email is invalid. Silence these errors
+                // to avoid spamming sentry.
+                Log::error('Paypal expired email not sent: ' . $e->getMessage());
+            }
+        }
 
         UserLogger::user($user)->log(UserAction::subPaypalExpiringWarning);
     }
diff --git a/app/Jobs/Emails/Subscriptions/UpcomingYearlyAlert.php b/app/Jobs/Emails/Subscriptions/UpcomingYearlyAlert.php
index 4707fc26a..80a02f0d7 100644
--- a/app/Jobs/Emails/Subscriptions/UpcomingYearlyAlert.php
+++ b/app/Jobs/Emails/Subscriptions/UpcomingYearlyAlert.php
@@ -42,11 +42,26 @@ class UpcomingYearlyAlert implements ShouldQueue
         }
 
         // Send an email to the user
-        Mail::to($user->email)
-            ->locale($user->locale)
-            ->send(
-                new UpcomingYearlyEmail($user)
-            );
+        if ( config('app.email_enabled') )
+        {
+            try {
+              Mail::to($user->email)
+                  ->locale($user->locale)
+                  ->send(
+                      new UpcomingYearlyEmail($user)
+                  );
+            } catch (\GuzzleHttp\Exception\ServerException $e) {
+                // Silence
+            } catch (TransportExceptionInterface $e) {
+                // Mail isn't configured or the server is unreachable. This is allowed, so don't rethrow.
+                Log::warning('Upcoming yearly renewal warning email not sent: ' . $e->getMessage());
+            } catch (Exception $e) {
+                // Something went wrong with mailgun, or the email is invalid. Silence these errors
+                // to avoid spamming sentry.
+                Log::error('Upcoming yearly renewal warning email not sent: ' . $e->getMessage());
+            }
+        }
+
         UserLogger::user($user)->log(UserAction::yearlyRenewWarning);
     }
 }
diff --git a/app/Jobs/Emails/Subscriptions/WelcomeSubscriptionEmailJob.php b/app/Jobs/Emails/Subscriptions/WelcomeSubscriptionEmailJob.php
index d8519dee1..3142fd44f 100644
--- a/app/Jobs/Emails/Subscriptions/WelcomeSubscriptionEmailJob.php
+++ b/app/Jobs/Emails/Subscriptions/WelcomeSubscriptionEmailJob.php
@@ -46,9 +46,24 @@ class WelcomeSubscriptionEmailJob implements ShouldQueue
 
         $tier = Tier::find($this->tierId);
 
-        Mail::to($user->email)
-            ->send(
-                new NewSubscriberMail($user, $tier)
-            );
+        if ( config('app.email_enabled') )
+        {
+            try {
+              Mail::to($user->email)
+                  ->send(
+                      new NewSubscriberMail($user, $tier)
+                  );
+            } catch (\GuzzleHttp\Exception\ServerException $e) {
+                // Silence
+            } catch (TransportExceptionInterface $e) {
+                // Mail isn't configured or the server is unreachable. This is allowed, so don't rethrow.
+                Log::debug('New subscription email not sent: ' . $e->getMessage());
+            } catch (Exception $e) {
+                // Something went wrong with mailgun, or the email is invalid. Silence these errors
+                // to avoid spamming sentry.
+                Log::error('New subscription email not sent: ' . $e->getMessage());
+            }
+        }
+
     }
 }
diff --git a/app/Jobs/Emails/WelcomeEmailJob.php b/app/Jobs/Emails/WelcomeEmailJob.php
index 07ef5c7aa..5e7627a84 100644
--- a/app/Jobs/Emails/WelcomeEmailJob.php
+++ b/app/Jobs/Emails/WelcomeEmailJob.php
@@ -54,18 +54,24 @@ class WelcomeEmailJob implements ShouldQueue
         }
         Log::info('WelcomeEmailJob', ['user' => $this->userId]);
 
-        // try {
-        Mail::to($user->email)
-            ->locale($this->language)
-            ->send(
-                new WelcomeEmail($user)
-            );
-        /*} catch (\GuzzleHttp\Exception\ServerException $e) {
-            // Silence
-        } catch (Exception $e) {
-            // Something went wrong with mailgun, or the email is invalid. Silence these errors
-            // to avoid spamming sentry.
-            throw $e;
-        }*/
+        if ( config('app.email_enabled') )
+        {
+            try {
+                Mail::to($user->email)
+                    ->locale($this->language)
+                    ->send(
+                        new WelcomeEmail($user)
+                    );
+            } catch (\GuzzleHttp\Exception\ServerException $e) {
+                // Silence
+            } catch (TransportExceptionInterface $e) {
+                // Mail isn't configured or the server is unreachable. This is allowed, so don't rethrow.
+                Log::debug('Welcome email not sent: ' . $e->getMessage());
+            } catch (Exception $e) {
+                // Something went wrong with mailgun, or the email is invalid. Silence these errors
+                // to avoid spamming sentry.
+                Log::error('Welcome email not sent: ' . $e->getMessage());
+            }
+        }
     }
 }
diff --git a/app/Jobs/Users/NewPassword.php b/app/Jobs/Users/NewPassword.php
index f9ca85a1e..af7343fee 100644
--- a/app/Jobs/Users/NewPassword.php
+++ b/app/Jobs/Users/NewPassword.php
@@ -37,19 +37,25 @@ class NewPassword implements ShouldQueue
             return;
         }
 
-        $target = app()->isProduction() ? $user->email : config('mail.from.address');
-        try {
-            Mail::to($target)
-                ->locale($user->locale ?? 'en-US')
-                ->send(
-                    new \App\Mail\Users\NewPassword($user)
-                );
-        } catch (ServerException $e) {
-            // Silence
-        } catch (Exception $e) {
-            // Something went wrong with mailgun, or the email is invalid. Silence these errors
-            // to avoid spamming sentry.
-            throw $e;
+        if ( config('app.email_enabled') )
+        {
+            $target = app()->isProduction() ? $user->email : config('mail.from.address');
+            try {
+                Mail::to($target)
+                    ->locale($user->locale ?? 'en-US')
+                    ->send(
+                        new \App\Mail\Users\NewPassword($user)
+                    );
+            } catch (ServerException $e) {
+                // Silence
+            } catch (TransportExceptionInterface $e) {
+                // Mail isn't configured or the server is unreachable. This is allowed, so don't rethrow.
+                Log::error('Change password email not sent: ' . $e->getMessage());
+            } catch (Exception $e) {
+                // Something went wrong with mailgun, or the email is invalid. Silence these errors
+                // to avoid spamming sentry.
+                Log::error('Change password email not sent: ' . $e->getMessage());
+            }
         }
     }
 }
diff --git a/config/app.php b/config/app.php
index b713d6960..40c8d3ad8 100644
--- a/config/app.php
+++ b/config/app.php
@@ -119,6 +119,7 @@ return [
 
     'name' => env('APP_NAME', 'Miscellany'),
     'email' => env('APP_EMAIL', 'you@example.com'),
+    'email_enabled' => env('APP_EMAIL_ENABLED', true),
 
     /*
     |--------------------------------------------------------------------------
-- 
2.52.0

